lib/matplotlib/backend_bases.py

```diff
@@ -175,20 +175,28 @@ def draw_if_interactive(cls):
                 cls.trigger_manager_draw(manager)
 
     @classmethod
-    def show(cls, block=None):
+    def show(cls, block=None, warn=True):
         """Show all figures.
 
         `show` blocks by calling `mainloop` if *block* is ``True``, or if it
         is ``None`` and we are neither in IPython's ``%pylab`` mode, nor in
         `interactive` mode.
         """
-        if cls.mainloop is None:
-            return
         managers = Gcf.get_all_fig_managers()
         if not managers:
             return
         for manager in managers:
-            manager.show()
+            try:
+                manager.show()
+            except NonGuiException:
+                if warn:
+                    import warnings
+                    warnings.warn(
+                        "matplotlib is currently using a non-GUI backend, "
+                        "so cannot show the figure")
```

Comment:
Might be nice to say which backend is being used

Reply:
Also, what's the policy now for warnings versus logging warnings?

Reply:
According to the [logging howto](https://docs.python.org/3/howto/logging.html)

> warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning
>
> logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted

The former is the case here. One should not try to call show (or expect that it does something) on a non-GUI backend. So, warnings.warn is correct.

Though, I don't think we need the `warn` parameter. Client applications can use the filter mechanism of the warnings module, if they really want to suppress the warning.